### PR TITLE
Cache definitions

### PR DESCRIPTION
**Pull request by HenningNormann**
## Description
Avoid reading process definition from file for each invocation

## Related Issue(s)
- #{93} in the platform team

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.App.Core/Internal/Process/ProcessReader.cs [hunk 1]
```diff
@@ -19,10 +20,18 @@
...
+            lock(_lockObject)
+            {
+                if (_definitions == null)
+                {
+                    XmlSerializer serializer = new XmlSerializer(typeof(Definitions));
+                    Definitions? definitions = (Definitions?)serializer.Deserialize(processService.GetProcessDefinition());
+                    _definitions = definitions ?? throw new InvalidOperationException("Failed to deserialize BPMN definitions. Definitions was null");
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tjololo) [reply src/Altinn.App.Core/Internal/Process/ProcessReader.cs]: Please fix this error

**Review by tjololo — CHANGES_REQUESTED** [review]
Please have a look at the codeQL error and make the tests pass

**Comment by tjololo** [comment]
@HenningNormann is there any reason you are choosing static fields over MemoryCache?

**Comment by HenningNormann** [comment]
> 



> @HenningNormann is there any reason you are choosing static fields over MemoryCache?

@tjololo To avoid changing a lot of interfaces to transfer the cache from a DI constructor to ProcessReader and because the value should be cached for the lifetime of the pod - without the need for any cache configuration.

**Comment by tjololo** [comment]
Has the time it takes to read this from the filesystem in a docker container been identified as a critical issue? We should not consume more memory than necessary. All applications in a cluster consume memory from the same pool and we should make sure that inactive apps don't consume more than necessary to ensure that apps that are receiving load have as much as possible resources available.

**Comment by HenningNormann** [comment]
> 

Yes, it has been identified as a significant resource consumer. The memory size of the cached object looks very small to me.

**Comment by tjololo** [comment]
Getting some fresh eyes on this after the holiday I think a simpler and better solution would be to leave the ProcessReader implementation as is and register it with the dependency injection as a singleton. I think this should yield the same performance boost as this solution, but free us from handling dirtyreads with locking in our code. Have a look at [ServiceCollectionExtension line 219](https://github.com/Altinn/app-lib-dotnet/blob/main/src/Altinn.App.Core/Extensions/ServiceCollectionExtensions.cs#L219)

**Comment by HenningNormann** [comment]
> Getting some fresh eyes on this after the holiday I think a simpler and better solution would be to leave the ProcessReader implementation as is and register it with the dependency injection as a singleton. I think this should yield the same performance boost as this solution, but free us from handling dirtyreads with locking in our code. Have a look at [ServiceCollectionExtension line 219](https://github.com/Altinn/app-lib-dotnet/blob/main/src/Altinn.App.Core/Extensions/ServiceCollectionExtensions.cs#L219)

Have changed as suggested

**Review by tjololo — APPROVED** [review]
:godmode:

**Comment by tjololo** [comment]
I will release a preview version (7.2.1-preview.1) so you can test that the changes have the desired affect

**Pull request merged**